packages/aws-cdk/test/integ/cli/cdk-helpers.ts

```diff
@@ -2,8 +2,99 @@ import * as child_process from 'child_process';
 import * as fs from 'fs';
 import * as os from 'os';
 import * as path from 'path';
-import { outputFromStack, testEnv, AwsClients } from './aws-helpers';
-import { TestEnvironment } from './test-helpers';
+import { outputFromStack, AwsClients } from './aws-helpers';
+import { ResourcePool, ILease } from './resource-pool';
+import { TestContext } from './test-helpers';
+
+const REGIONS = process.env.AWS_REGIONS
+  ? process.env.AWS_REGIONS.split(',')
+  : [process.env.AWS_REGION ?? process.env.AWS_DEFAULT_REGION ?? 'us-east-1'];
+
+process.stdout.write(`Using regions: ${REGIONS}\n`);
+
+const REGION_POOL = new ResourcePool(REGIONS);
+
+
+export type AwsContext = { readonly aws: AwsClients };
+
+/**
+ * Higher order function to execute a block with an AWS client setup
+ *
+ * Allocate the next region from the REGION pool and dispose it afterwards.
+ */
+export function withAws<A extends TestContext>(block: (context: A & AwsContext) => Promise<void>) {
+  return (context: A) => REGION_POOL.using(async (region) => {
+    const aws = await AwsClients.forRegion(region, context.output);
+    await sanityCheck(aws);
+
+    return block({ ...context, aws });
+  });
+}
+
+/**
+ * Higher order function to execute a block with a CDK app fixture
+ *
+ * Requires an AWS client to be passed in.
+ *
+ * For backwards compatibility with existing tests (so we don't have to change
+ * too much) the inner block is expecte to take a `TestFixture` object.
+ */
+export function withCdkApp<A extends TestContext & AwsContext>(block: (context: TestFixture) => Promise<void>) {
+  return async (context: A) => {
+    const randy = randomString();
+    const stackNamePrefix = `cdktest-${randy}`;
+    const integTestDir = path.join(os.tmpdir(), `cdk-integ-${randy}`);
+    const regionLease = await REGION_POOL.take();
+
+    context.output.write(` Stack prefix:   ${stackNamePrefix}\n`);
+    context.output.write(` Test directory: ${integTestDir}\n`);
+    context.output.write(` Region:         ${regionLease.value}\n`);
+
+    await cloneDirectory(path.join(__dirname, 'app'), integTestDir, context.output);
+    const fixture = new TestFixture(
+      integTestDir,
+      stackNamePrefix,
+      context.output,
+      regionLease,
+      context.aws);
```

Comment:
Why take the lease here if it's going to be released within the `dispose()` method? As mentioned previously, would be nice to see `take()` and `dispose()` in the same block/scope, making it easier to reason about.

Would it make more sense to take the least as part of the `TextFixture` constructor?

Reply:
Oh wait. You are correct that i left a bug.

Reply:
> Would it make more sense to take the least as part of the `TextFixture` constructor?

Can't. It's async.